packages/router/src/grpc/offscreen-client.ts

```diff
@@ -0,0 +1,71 @@
+import { WitnessAndBuildRequest } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1alpha1/view_pb';
+import {
+  WitnessData,
+  Action,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/transaction/v1alpha1/transaction_pb';
+import { InternalRequest, InternalResponse } from '@penumbra-zone/types/src/internal-msg/shared';
+import {
+  ActionBuildMessage,
+  ActionBuildRequest,
+  OffscreenMessage,
+} from '@penumbra-zone/types/src/internal-msg/offscreen';
+
+const OFFSCREEN_DOCUMENT_PATH = '/offscreen.html';
+
+let active = 0;
+
+const activateOffscreen = async () => {
+  // @ts-expect-error: no types available yet
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
+  const getContexts = chrome.runtime.getContexts as (x: unknown) => Promise<unknown[]>;
+  const offscreenExists = getContexts({
+    contextTypes: ['OFFSCREEN_DOCUMENT'],
+    documentUrls: [OFFSCREEN_DOCUMENT_PATH],
+  }).then(contexts => contexts.length > 0);
+
+  if (active && (await offscreenExists)) active++;
+  else {
+    active++;
+    await chrome.offscreen.createDocument({
+      url: OFFSCREEN_DOCUMENT_PATH,
+      reasons: [chrome.offscreen.Reason.WORKERS],
+      justification: 'Manages Penumbra transaction WASM workers',
+    });
+  }
+};
+
+const releaseOffscreen = () => {
+  active--;
+  if (!active) void chrome.offscreen.closeDocument();
+};
+
+const sendOffscreenMessage = async <T extends OffscreenMessage>(
+  req: InternalRequest<T>,
+): Promise<InternalResponse<ActionBuildMessage>> => {
+  try {
+    await activateOffscreen();
+    return await chrome.runtime.sendMessage<InternalRequest<T>, InternalResponse<T>>(req);
+  } finally {
+    releaseOffscreen();
```

Comment:
Does this need to be a void function given the surrounding environment is async? We could await this. Is it a void so the release doesn't slow down the return of the response?

Reply:
It doesn't seem to affect the build time, but I don't see a reason to make the offscreen cleanup a synchronous operation by calling await. Or maybe it doesn't make a difference between the finally block has to be called regardless before the function returns.

Reply:
The offscreen API documentation mentions `await chrome.offscreen.closeDocument()`, so we should do the same. I changed it to await.